public/app/features/dashboard-scene/scene/AddLibraryPanelWidget.tsx

```diff
@@ -0,0 +1,184 @@
+import { css, cx, keyframes } from '@emotion/css';
+import React from 'react';
+import tinycolor from 'tinycolor2';
+
+import { GrafanaTheme2 } from '@grafana/data';
+import {
+  SceneComponentProps,
+  SceneGridItem,
+  SceneGridLayout,
+  SceneGridRow,
+  SceneObjectBase,
+  SceneObjectState,
+} from '@grafana/scenes';
+import { LibraryPanel } from '@grafana/schema';
+import { IconButton, useStyles2 } from '@grafana/ui';
+import { Trans } from 'app/core/internationalization';
+import {
+  LibraryPanelsSearch,
+  LibraryPanelsSearchVariant,
+} from 'app/features/library-panels/components/LibraryPanelsSearch/LibraryPanelsSearch';
+
+import { getDashboardSceneFor } from '../utils/utils';
+
+import { DashboardScene } from './DashboardScene';
+import { LibraryVizPanel } from './LibraryVizPanel';
+
+export interface AddLibraryPanelWidgetState extends SceneObjectState {
+  key: string;
+}
+
+export class AddLibraryPanelWidget extends SceneObjectBase<AddLibraryPanelWidgetState> {
+  public constructor(state: AddLibraryPanelWidgetState) {
+    super({
+      ...state,
+    });
+  }
+
+  private get _dashboard(): DashboardScene {
+    return getDashboardSceneFor(this);
+  }
+
+  public getDashboard(): DashboardScene {
+    return this._dashboard;
+  }
+
+  public onCancelAddPanel = (evt: React.MouseEvent<HTMLButtonElement>) => {
+    evt.preventDefault();
+
+    if (!(this._dashboard.state.body instanceof SceneGridLayout)) {
+      throw new Error('Trying to remove the library panel widget in a layout that is not SceneGridLayout');
+    }
+
+    const sceneGridLayout = this._dashboard.state.body;
+    const children = [];
+
+    for (const child of sceneGridLayout.state.children) {
+      if (child.state.key !== this.parent?.state.key) {
+        children.push(child);
+      }
+
+      if (child instanceof SceneGridRow) {
+        const rowChildren = [];
+
+        for (const rowChild of child.state.children) {
+          if (rowChild instanceof SceneGridItem && rowChild.state.key !== this.parent?.state.key) {
+            rowChildren.push(rowChild);
+          }
+        }
+
+        child.setState({ children: rowChildren });
+      }
+    }
+
+    sceneGridLayout.setState({ children });
+  };
+
+  public onAddLibraryPanel = (panelInfo: LibraryPanel) => {
+    if (!(this._dashboard.state.body instanceof SceneGridLayout)) {
+      throw new Error('Trying to add a library panel in a layout that is not SceneGridLayout');
+    }
+
+    const sceneGridLayout = this._dashboard.state.body;
+
+    const body = new LibraryVizPanel({
+      title: 'Panel Title',
+      uid: panelInfo.uid,
+      name: panelInfo.name,
+      //TODO modify this to be panelKey after https://github.com/grafana/grafana/pull/83420 gets merged
+      key: this.state.key,
+    });
+
+    for (const child of sceneGridLayout.state.children) {
+      if (child instanceof SceneGridItem && child.state.key === this.parent?.state.key) {
+        child.setState({ body });
+        return;
```

Comment:
Do we really need this search? Isn't this.parent the same as child? Then meybe we can do `this.parent.setState({body})` ?

Could be that I'm missing something here.

Reply:
That's true! Thanks for catching that, makes the code so much better. Pushing a commit soon!